### DISCUSSION TITLE
Updating worspace incremental refresh model

### DISCUSSION THREAD

**Discussion by 10xape** [TabularEditor/TabularEditor3#1416] [Data Modeling]
Hi,

I'm trying to understand how TabularEditor 3 handles workspace incremental model changes in the following scenarios:

Multiple tables with multiple relations 

Adding new measures
Adding or editing relationships
Adding columns or calculations to existing tables
Modifying source expressions in M code
Adding, removing tables

I've reviewed the documentation but would appreciate a simple, clear explanation and step-by-step instructions for how to make these changes take effect across all date partitions.

Could you please break this down in straightforward grade 6 terms?

Thanks,
Matt

**Comment by 10xape** [comment]
And also what changes may affect the scheduled refresh, I have added table and now have the following error popping out during the scheduled refresh 
![image](https://github.com/user-attachments/assets/5cdcce99-ebf2-4b6b-8aa3-31142b8bae03)

**Reply by otykier** [reply]
Scheduled refreshes will generally not be impacted by changes, but of course if you introduce an error in, say, a source expression, then the scheduled refresh will not be able to run. Always ensure that refresh still works by performing a manual refresh (either through the service or through TE3) after making changes to source expressions. Ideally, this should be done in a dev/test workspace, so as not to impact any scheduled refreshes you might have in production.

**Comment by otykier** [comment]
Hi Matt,

The confusion probably arises from the fact that Power BI Desktop performs refreshes automatically as needed, where as with Tabular Editor, it’s your responsibility as the model developer to perform the refreshes.

There are different types of refreshes that can be performed. When you initiate such a refresh, Tabular Editor simply sends the corresponding “refresh” command to the server, which performs the actual refresh while continuously sending progress events to Tabular Editor

- **Data refresh**: Loads data from the source into the specified tables or partitions. Doesn’t evaluate calculated columns and calculated tables, and doesn’t bring relationships into a queryable state.
- **Calc refresh**: Does *not* load data from any sources, but recalculates calculated columns, tables and relationships. Also necessary if you add or modify a calculation group.
- **Full refresh**: Corresponds to **Data refresh** on all tables/partitions followed by **Calc refresh**
- **Auto refresh**: Same as **Full refresh** but only loads data to tables/partitions that do not already contain data.

More information here: https://learn.microsoft.com/en-us/analysis-services/tmsl/refresh-command-tmsl?view=asallproducts-allversions

As to your questions:

- **Adding new measures**: Can be done at any time. Doesn’t require any type of refresh.
- **Adding or editing relationships**: Requires a calc refresh
- **Adding columns or calculations to existing tables**: Adding data columns requires a *full* refresh of the table, before the new columns can be used in visuals. Adding/modifying calc columns requires a calc refresh.
- **Modifying source expressions in M code**: Doesn’t affect data that’s already loaded to the model but a data refresh is needed to load data using the modified source expressions.
- **Adding, removing tables**: Can be done at any time, and may require a calc refresh before the model can be queried. There will not be any data in new tables until a data refresh has been performed.

Hope this helps.
